packages/application-shell-connectors/src/components/application-state/application-state.js

```diff
@@ -0,0 +1,98 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { wrapDisplayName } from 'recompose';
+import moment from 'moment-timezone';
+import omit from 'lodash.omit';
+
+const { Provider, Consumer } = React.createContext({});
+
+const defaultTimeZone = moment.tz.guess() || 'Etc/UTC';
+
+// Expose only certain fields as some of them are only meant to
+// be used internally in the AppShell
+const mapUserPropsToApplicationState = user => ({
+  id: user.id,
+  email: user.email,
+  firstName: user.firstName,
+  lastName: user.lastName,
+  // NOTE: this is an alias for the original field `user.language`.
+  // This field is used for the application intl locale though, therefore
+  // we might want to expose it with a more explicit name to avoid
+  // confusion.
+  applicationLocale: user.language,
+  numberFormat: user.numberFormat,
+  timeZone: user.timeZone || defaultTimeZone,
+});
+
+// Expose only certain fields as some of them are only meant to
+// be used internally in the AppShell
+const mapProjectPropsToApplicationState = (project, projectDataLocale) => ({
+  key: project.key,
+  version: project.version,
+  name: project.name,
+  countries: project.countries,
+  currencies: project.currencies,
+  languages: project.languages,
+  permissions: omit(project.permissions, ['__typename']),
+  // Additional fields that do not belong directly to a project
+  ...(projectDataLocale ? { dataLocale: projectDataLocale } : {}),
+});
+
+const ApplicationStateProvider = props => (
+  <Provider
+    value={{
```

Comment:
I would suggest to memoize this. It's often said to be a good practice. Also this whole spreading down there makes me feel a bit unease given how far up the component tree we are.